File: internal/handler/definitions.go
package handler

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"
	"github.com/marekvalenta/inventory-management/internal/service"
)

type DefinitionHandler struct {
	svc *service.DefinitionService
}

func NewDefinitionHandler(svc *service.DefinitionService) *DefinitionHandler {
	return &DefinitionHandler{svc: svc}
}

type CreateDefinitionRequest struct {
	Name        *string                    `json:"name" validate:"required"`
	Description *string                    `json:"description"`
	ParentDefID *string                    `json:"parent_def_id"`
	Unit        *string                    `json:"unit"`
	IsContainer *bool                       `json:"is_container"`
	Fields      []CreateFieldRequest       `json:"fields"`
	TagIDs      []string                   `json:"tag_ids"`
}

type CreateFieldRequest struct {
	FieldName       string           `json:"field_name" validate:"required"`
	FieldType       string           `json:"field_type" validate:"required,oneof=text number boolean date enum"`
	EnumValues      *json.RawMessage `json:"enum_values"`
	IsRequired      bool             `json:"is_required"`
	DisplayOrder    int              `json:"display_order"`
	DefaultValue    *string          `json:"default_value"`
	IsChildEditable bool             `json:"is_child_editable"`
}

type UpdateDefinitionRequest struct {
	Name        *string             `json:"name"`
	Description *string             `json:"description"`
	ParentDefID *string             `json:"parent_def_id"`
	Unit        *string             `json:"unit"`
	IsContainer *bool                `json:"is_container"`
	Fields      *[]service.CreateFieldInput `json:"fields"`
	TagIDs      *[]string           `json:"tag_ids"`
}

type OverrideRequest struct {
	ParentFieldID string  `json:"parent_field_id"`
	DefaultValue  *string `json:"default_value"`
}

type UpdateOverridesRequest struct {
	Overrides []OverrideRequest `json:"overrides"`
}

func (h *DefinitionHandler) List(w http.ResponseWriter, r *http.Request) {
	defs, err := h.svc.GetAll()
	if err != nil {
		RespondWithError(w, err)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(defs)
}

func (h *DefinitionHandler) Get(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")

	def, err := h.svc.GetByID(id)
	if err != nil {
		RespondWithError(w, err)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(def)
}

func (h *DefinitionHandler) Create(w http.ResponseWriter, r *http.Request) {
	var req CreateDefinitionRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		RespondWithError(w, service.ErrInvalidInput)
		return
	}

	if req.Name == nil {
		RespondWithError(w, service.ErrInvalidInput)
		return
	}

	if req.Fields == nil {
		req.Fields = []CreateFieldRequest{}
	}

	fieldInputs := make([]service.CreateFieldInput, len(req.Fields))
	for i, f := range req.Fields {
		fieldInputs[i] = service.CreateFieldInput{
			FieldName:       f.FieldName,
			FieldType:       f.FieldType,
			EnumValues:      f.EnumValues,
			IsRequired:      f.IsRequired,
			DisplayOrder:    f.DisplayOrder,
			DefaultValue:    f.DefaultValue,
			IsChildEditable: f.IsChildEditable,
		}
	}

	def, err := h.svc.Create(service.CreateDefinitionInput{
		Name:        *req.Name,
		Description: req.Description,
		ParentDefID: req.ParentDefID,
		Unit:        req.Unit,
		IsContainer: req.IsContainer,
		Fields:      fieldInputs,
		TagIDs:      req.TagIDs,
	})
	if err != nil {
		if strings.Contains(err.Error(), "already exists") {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusConflict)
			json.NewEncoder(w).Encode(ErrorResponse{
				Error: err.Error(),
				Code:  "duplicate_name",
			})
			return
		}
		RespondWithError(w, err)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(def)
}

func (h *DefinitionHandler) Update(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")

	var req UpdateDefinitionRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		RespondWithError(w, service.ErrInvalidInput)
		return
	}

	def, err := h.svc.Update(id, service.UpdateDefinitionInput{
		Name:        req.Name,
		Description: req.Description,
		ParentDefID: req.ParentDefID,
		Unit:        req.Unit,
		IsContainer: req.IsContainer,
		Fields:      req.Fields,
		TagIDs:      req.TagIDs,
	})
	if err != nil {
		if strings.Contains(err.Error(), "already exists") {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusConflict)
			json.NewEncoder(w).Encode(ErrorResponse{
				Error: err.Error(),
				Code:  "duplicate_name",
			})
			return
		}
		RespondWithError(w, err)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(def)
}

func (h *DefinitionHandler) Delete(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")

	err := h.svc.Delete(id)
	if err != nil {
		if strings.Contains(err.Error(), "child definitions") || strings.Contains(err.Error(), "item instances") {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusConflict)
			json.NewEncoder(w).Encode(ErrorResponse{
				Error: err.Error(),
				Code:  "conflict",
			})
			return
		}
		RespondWithError(w, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func (h *DefinitionHandler) UpdateOverrides(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")

	var req UpdateOverridesRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		RespondWithError(w, service.ErrInvalidInput)
		return
	}

	overrides := make([]service.OverrideInput, len(req.Overrides))
	for i, o := range req.Overrides {
		overrides[i] = service.OverrideInput{
			ParentFieldID: o.ParentFieldID,
			DefaultValue:  o.DefaultValue,
		}
	}

	result, err := h.svc.UpdateOverrides(id, overrides)
	if err != nil {
		RespondWithError(w, err)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"overrides": result,
	})
}

func (h *DefinitionHandler) RegisterRoutes(r chi.Router) {
	r.Get("/definitions", h.List)
	r.Get("/definitions/{id}", h.Get)
	r.Post("/definitions", h.Create)
	r.Put("/definitions/{id}", h.Update)
	r.Delete("/definitions/{id}", h.Delete)
	r.Put("/definitions/{id}/overrides", h.UpdateOverrides)
}

File: frontend/src/api/definitions.ts
import { apiFetch } from './client'

export interface DefinitionSummary {
  id: string
  name: string
  description: string | null
  parent_def_id: string | null
  parent_def_name: string | null
  unit: string | null
  is_container: boolean
  total_instances: number
  tags: DefinitionTag[]
  created_at: string
  updated_at: string
}

export interface DefinitionTag {
  id: string
  name: string
  color: string | null
  created_at: string
  updated_at: string
}

export interface DefinitionField {
  id: string
  field_name: string
  field_type: 'text' | 'number' | 'boolean' | 'date' | 'enum'
  enum_values: string[] | null
  is_required: boolean
  display_order: number
  default_value: string | null
  is_child_editable: boolean
  inherited_from_def_id: string | null
}

export interface LocationInstanceCount {
  location_id: string
  location_name: string
  instance_count: number
  total_quantity: number
}

export interface ParentInstanceCount {
  parent_instance_id: string
  parent_instance_name: string
  location_id: string
  location_name: string
  instance_count: number
  total_quantity: number
}

export interface InstancesSummary {
  total_instances: number
  total_quantity: number
  by_location: LocationInstanceCount[]
  by_parent_instance: ParentInstanceCount[]
}

export interface DefinitionDetail {
  id: string
  name: string
  description: string | null
  parent_def_id: string | null
  parent_def_name: string | null
  unit: string | null
  is_container: boolean
  created_at: string
  updated_at: string
  fields: DefinitionField[]
  tags: DefinitionTag[]
  instances_summary: InstancesSummary
  child_definition_count: number
}

export interface CreateFieldInput {
  field_name: string
  field_type: 'text' | 'number' | 'boolean' | 'date' | 'enum'
  enum_values: string[] | null
  is_required: boolean
  display_order: number
  default_value: string | null
  is_child_editable: boolean
}

export interface CreateDefinitionRequest {
  name: string
  description?: string | null
  parent_def_id?: string | null
  unit?: string | null
  is_container?: boolean
  fields?: CreateFieldInput[]
  tag_ids?: string[]
}

export interface UpdateDefinitionRequest {
  name?: string
  description?: string | null
  parent_def_id?: string | null
  unit?: string | null
  is_container?: boolean
  fields?: CreateFieldInput[]
  tag_ids?: string[]
}

export interface OverrideInput {
  parent_field_id: string
  default_value: string | null
}

export interface OverrideResponse {
  definition_id: string
  parent_field_id: string
  default_value: string | null
}

export function fetchDefinitions(): Promise<DefinitionSummary[]> {
  return apiFetch<DefinitionSummary[]>('/api/v1/definitions')
}

export function fetchDefinition(id: string): Promise<DefinitionDetail> {
  return apiFetch<DefinitionDetail>(`/api/v1/definitions/${encodeURIComponent(id)}`)
}

export function createDefinition(data: CreateDefinitionRequest): Promise<DefinitionDetail> {
  return apiFetch<DefinitionDetail>('/api/v1/definitions', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function updateDefinition(id: string, data: UpdateDefinitionRequest): Promise<DefinitionDetail> {
  return apiFetch<DefinitionDetail>(`/api/v1/definitions/${encodeURIComponent(id)}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  })
}

export function deleteDefinition(id: string): Promise<void> {
  return apiFetch<void>(`/api/v1/definitions/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
}

export function updateOverrides(id: string, overrides: OverrideInput[]): Promise<{ overrides: OverrideResponse[] }> {
  return apiFetch<{ overrides: OverrideResponse[] }>(
    `/api/v1/definitions/${encodeURIComponent(id)}/overrides`,
    {
      method: 'PUT',
      body: JSON.stringify({ overrides }),
    },
  )
}
